File: pkg/java/src/main/java/dev/openfga/language/OpenFgaDslErrorListener.java
package dev.openfga.language;

import static java.util.Collections.unmodifiableList;

import dev.openfga.language.errors.ErrorProperties;
import dev.openfga.language.errors.Metadata;
import dev.openfga.language.errors.StartEnd;
import dev.openfga.language.errors.SyntaxError;
import java.util.ArrayList;
import java.util.BitSet;
import java.util.List;
import org.antlr.v4.runtime.*;
import org.antlr.v4.runtime.atn.ATNConfigSet;
import org.antlr.v4.runtime.dfa.DFA;

public class OpenFgaDslErrorListener implements ANTLRErrorListener {

    private final List<SyntaxError> errors = new ArrayList<>();

    public List<SyntaxError> getErrors() {
        return unmodifiableList(errors);
    }

    @Override
    // line is one based, i.e. the first line will be 1
    // column is zero based, i.e. the first column will 0
    public void syntaxError(
            Recognizer<?, ?> recognizer,
            Object offendingSymbol,
            int line,
            int column,
            String message,
            RecognitionException e) {
        Metadata metadata = null;
        var columnOffset = 0;

        if (offendingSymbol instanceof Token) {
            metadata = new Metadata(((Token) offendingSymbol).getText());
            columnOffset = metadata.getSymbol().length();
        }

        var properties =
                new ErrorProperties(new StartEnd(line, line), new StartEnd(column, column + columnOffset), message);
        this.errors.add(new SyntaxError(properties, metadata, e));
    }

    @Override
    public void reportAmbiguity(
            Parser parser, DFA dfa, int i, int i1, boolean b, BitSet bitSet, ATNConfigSet atnConfigSet) {}

    @Override
    public void reportAttemptingFullContext(
            Parser parser, DFA dfa, int i, int i1, BitSet bitSet, ATNConfigSet atnConfigSet) {}

    @Override
    public void reportContextSensitivity(Parser parser, DFA dfa, int i, int i1, int i2, ATNConfigSet atnConfigSet) {}
}

File: pkg/csharp/src/OpenFga.Language/OpenFgaDslErrorListener.cs
using Antlr4.Runtime;
using OpenFga.Language.Errors;

namespace OpenFga.Language;

/// <summary>
/// Error listener for OpenFGA DSL parsing that collects syntax errors during lexer and parser operations.
/// Implements ANTLR error listener interfaces to capture and store parsing errors.
/// </summary>
public class OpenFgaDslErrorListener : IAntlrErrorListener<int>, IAntlrErrorListener<IToken> {
    private readonly List<SyntaxError> _errors = new();

    /// <summary>
    /// Gets the read-only list of syntax errors collected during parsing.
    /// </summary>
    public IReadOnlyList<SyntaxError> Errors => _errors.AsReadOnly();

    /// <summary>
    /// Handles syntax errors from the lexer (token recognition errors).
    /// </summary>
    /// <param name="output">The output writer</param>
    /// <param name="recognizer">The recognizer that detected the error</param>
    /// <param name="offendingSymbol">The offending symbol</param>
    /// <param name="line">The line number where the error occurred</param>
    /// <param name="column">The column number where the error occurred</param>
    /// <param name="msg">The error message</param>
    /// <param name="e">The recognition exception</param>
    public void SyntaxError(TextWriter output, IRecognizer recognizer, int offendingSymbol, int line, int column, string msg, RecognitionException e) {
        // For lexer errors
        var properties = new ErrorProperties(
            new StartEnd(line - 1, line - 1),
            new StartEnd(column, column),
            msg);

        _errors.Add(new SyntaxError(properties, null, e));
    }

    /// <summary>
    /// Handles syntax errors from the parser (grammar recognition errors).
    /// </summary>
    /// <param name="output">The output writer</param>
    /// <param name="recognizer">The recognizer that detected the error</param>
    /// <param name="offendingSymbol">The offending token</param>
    /// <param name="line">The line number where the error occurred</param>
    /// <param name="column">The column number where the error occurred</param>
    /// <param name="msg">The error message</param>
    /// <param name="e">The recognition exception</param>
    public void SyntaxError(TextWriter output, IRecognizer recognizer, IToken offendingSymbol, int line, int column, string msg, RecognitionException e) {
        // For parser errors
        // line is one based, i.e. the first line will be 1
        // column is zero based, i.e. the first column will 0
        Metadata? metadata = null;
        var columnOffset = 0;

        if (offendingSymbol != null) {
            metadata = new Metadata(offendingSymbol.Text);
            columnOffset = metadata.Symbol.Length;
        }

        var properties = new ErrorProperties(
            new StartEnd(line - 1, line - 1),
            new StartEnd(column, column + columnOffset),
            msg);

        _errors.Add(new SyntaxError(properties, metadata, e));
    }
}